Implement regex matching for request path segments

ELD opened the pull request:
Adds anchored (^..$) regex matching to the request path segment handling
and drops the `unimplemented!()` macro in favor of a concrete
implementation.

This also adds an `OrderedRegex` newtype as the exist Regex didn't
implement the traits required for PartialEq, Eq, PartialOrd, and Ord.
See the following Github issues for reasons why Regex doesn't natively
implement this:
https://github.com/rust-lang/regex/issues/313
https://github.com/rust-lang/regex/issues/364

This (hopefully) addresses and closes #10.

Only concerns with this pull request are the uses of `unwrap` when constructing the regular expression. Not sure how we want to handle errors for that.

ELD commented:
I noticed that code coverage decreased likely as a result of adding a type with some manual trait implementations. I can write tests for that type to hopefully restore coverage, but the implementations are trivial. I also can write some docs for the type if they're so requested, but it's not intended to be public to the greater crate.

Let me know what you think of the code and if it could eventually be merged.

smangelsdorf commented:
Just doing some final checks on this, I've also noted that the `struct OrderedRegex` and its associated function are being warned about by the `missing_docs` lint. Would you mind please adding a brief doc comment for them?

ELD commented on src/router/tree/node.rs:
```diff
@@ -4,12 +4,49 @@ use std::cmp::Ordering;
+    fn partial_cmp(&self, other: &OrderedRegex) -> Option<Ordering> {
+        return if self.0.as_str() > other.0.as_str() {
+            Some(Ordering::Greater)
+        } else if self.0.as_str() < other.0.as_str() {
```
Yes, both can be simplified to that. It totally escaped my mind at the time. Thanks for catching that :)

ELD commented:
@smangelsdorf I went ahead and added the doc comment for the struct and associated function. I also changed the PartialOrd and Ord implementations in line with your comments. You were right that it could be implemented that way. It escaped my mind at the time I wrote the code but your solution made much more sense than mine.

Is it fine to keep them as two separate commits or would you like it to be squashed into one?

smangelsdorf commented:
Thanks for this @ELD. Great to land this feature!

Separate commits are fine, thanks.

ELD commented:
@smangelsdorf No problem! I'm happy to contribute to Gotham!
